### PR TITLE
Improve check changes only in O2 check code

### PR DESCRIPTION
**Pull request by dberzano**
Rationale: speed up automatic checks by checking only what effectively changed, and what was affected by such changes.

* By default it checks all files
* If `ALIBUILD_BASE_HASH` is given, checks current working area (including
  untracked files) with respect to that commit: useful for local tests
* If both `ALIBUILD_BASE_HASH` and `ALIBUILD_HEAD_HASH` are given, check only
  what changed between them: useful for centralized PR tests
* If more than 50 files changed, revert to checking all files: computing the
  affected files would be too expensive

### CODE REVIEW

**Comment by dberzano** [comment]
This change explicitly requires both commits to be specified and computes a diff list containing only the actual changes between the specified head and the PR base (the "three dots", `...`). [Here](https://stackoverflow.com/questions/462974/what-are-the-differences-between-double-dot-and-triple-dot-in-git-com) there is an excellent explanation (with Venn diagrams) to see what it does.

This goes together with alisw/ali-bot#464, where we also pass the correct "head hash" along with the PR "base hash".

@sawenzel do you agree?

**Comment by dberzano** [comment]
Just updated according to what we've discussed. Note that the variable naming has changed to avoid confusion between what is head and base.

File: o2checkcode.sh [hunk 1]
```diff
@@ -19,20 +19,21 @@ cp "${O2_ROOT}"/compile_commands.json .
...
+# changed files (including the untracked ones) if the list of relevant files that changed is up to
+# 50 entries long
+if [[ $ALIBUILD_UPSTREAM_HASH ]]; then
+  pushd "$O2_SRC"
+    ( git diff --name-only $ALIBUILD_UPSTREAM_HASH${ALIBUILD_PR_HASH:+...$ALIBUILD_PR_HASH} ; git ls-files --others --exclude-standard ) | grep -E '\.cxx$|\.h$' | sort -u > $BUILDDIR/changed
+    if [[ $(cat $BUILDDIR/changed | wc -l) -le 50 ]]; then
+      O2_CHECKCODE_CHANGEDFILES=$(while read FILE; do [[ -e "$O2_SRC/$FILE" ]] && echo "$FILE" || true; done < <(cat $BUILDDIR/changed) | \
+                                  xargs echo | sed -e 's/ /:/g')
+      if [[ ! $O2_CHECKCODE_CHANGEDFILES ]]; then
+        echo "Nothing changed with respect to base commit: not checking anything"
```
Comment (sawenzel) [inline o2checkcode.sh]: probably this message needs to be changed, too (since you exchanged HEAD<->BASE).

File: o2checkcode.sh [hunk 2]
```diff
@@ -19,20 +19,21 @@ cp "${O2_ROOT}"/compile_commands.json .
...
-                              while read FILE; do [[ -e "$O2_SRC/$FILE" ]] && echo "$FILE" || true; done | \
-                              xargs echo | sed -e 's/ /:/g')
-  if [[ ! $O2_CHECKCODE_CHANGEDFILES ]]; then
-    echo "Nothing changed with respect to base commit: not checking anything"
-    exit 0
-  fi
+# We have something to compare our working directory to (ALIBUILD_UPSTREAM_HASH). We check only the
+# changed files (including the untracked ones) if the list of relevant files that changed is up to
+# 50 entries long
+if [[ $ALIBUILD_UPSTREAM_HASH ]]; then
```
Comment (sawenzel) [inline o2checkcode.sh]: How can a user now use the recipe? Does she have to define `ALIBUILD_UPSTREAM_HASH`? In the old version it was quite clear that she had to define more meaningful variables `O2CHECKCODE_...`

Comment (dberzano) [inline o2checkcode.sh]: The rationale was to have a single variable to trigger the behaviour, but now that I am writing this comment I realize that:

* for central use cases, the recipe stays independent and should use `ALIBUILD_UPSTREAM_HASH`
* for local use cases, it's more intuitive to export `O2CHECKCODE...`

So... I think you are right. Let me update this.

File: o2checkcode.sh (hunk 1 above)
Comment (dberzano) [inline o2checkcode.sh]: You are right. Update coming.

**Comment by dberzano** [comment]
@sawenzel can you check whether the variable names `O2CHECKCODE_UPSTREAM_HASH` and `O2CHECKCODE_WORK_HASH` work for you? As said I would like to avoid the names "base" and "head".

(See, more than 5 minutes work...)

File: o2checkcode.sh [hunk 3]
```diff
@@ -19,26 +19,33 @@ cp "${O2_ROOT}"/compile_commands.json .
...
 
-BASECOMMIT=${O2CHECKCODE_BASECOMMIT:=${ALIBUILD_PR_BASE}}
-HEADCOMMIT=${O2CHECKCODE_HEADCOMMIT:=${ALIBUILD_PR_HEAD}}
+# We have something to compare our working directory to. We check only the
+# changed files (including the untracked ones) if the list of relevant files
+# that changed is up to 50 entries long. User can trigger a check by setting the
+# upstream reference hash in the O2CHECKCODE_UPSTREAM_HASH. User can also export
+# O2CHECKCODE_WORK_HASH if she does not want to compare the current head plus
+# the working directory.
+ALIBUILD_UPSTREAM_HASH=${ALIBUILD_UPSTREAM_HASH:=$O2CHECKCODE_UPSTREAM_HASH}
```
Comment (sawenzel) [inline o2checkcode.sh]: I find `UPSTREAM` still a bit confusing. (But I don't want to block this any longer either).

**Pull request merged**